### PR TITLE
feat: send message delivery confirmation (WPB-9023)

### PR DESCRIPTION
**Pull request by yamilmedina**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9023" title="WPB-9023" target="_blank">TaskWPB-9023</a>  [Android] Delivered status of messages not properly propagating
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Delivery confirmation of messages is not being delivered to other clients

### Causes (Optional)

Never implemented, we only implemented read receipts confirmations.

### Solutions

- While receiving messages, we enqueue it per conversation. 
- When sending, we introduced a conversations cache, _this is needed to check the conversation type_ since delivery confirmations needs to be sent only for 1 to 1 conversations.
- We debounce the collection, and sending so we don't spam the network with many requests and send in batches.
- As discussed with @vitorhugods and @alexandreferris we can introduce or make use of this cache at a larger scale for other parts of the app, but is something that we might do in a progressive adoption, for now only for this feature.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Have the app installed in the conversations list or in the background, if you receive a 1:1 conversation message, you should see the `Delivered` status, eg. on web.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 1]
```diff
@@ -1735,8 +1742,10 @@ class UserSessionScope internal constructor(
             this,
             userScopedLogger,
         )
-    val messages: MessageScope
-        get() = MessageScope(
+    }
+
+    val messages: MessageScope by lazy {
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1747]: We noticed while checking in the profiler,  that we were creating many instances of `MessageScope`.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 2]
```diff
@@ -474,6 +473,10 @@ class UserSessionScope internal constructor(
     private val logoutCallback: LogoutCallback,
 ) : CoroutineScope {
 
+    init {
+        onUserSessionScopeCreated()
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:477]: Small refactor to have better understanding (and kdocs) of what work we do subscribe when we create a session in the app.

**Review by Garzas — APPROVED** [review]
Cool improvements with new feature 🚀

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/conversation/ConversationDAO.kt [hunk 3]
```diff
@@ -132,6 +132,11 @@ interface ConversationDAO {
     suspend fun getEstablishedSelfMLSGroupId(): String?
 
     suspend fun selectGroupStatusMembersNamesAndHandles(groupID: String): EpochChangesDataEntity?
+
+    /**
+     * TODO: Remove later when implementing general conversation cache in the app.
+     */
+    suspend fun observeConversationDetailsById(conversationId: QualifiedIDEntity): Flow<ConversationViewEntity?>
```
Comment (vitorhugods) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/conversation/ConversationDAO.kt:139]: Just to confirm:
The idea of not caching right now is avoiding having to change tests in `ConversationDAOTest`, right?

Cause other than that, it is exactly like `observeGetConversationByQualifiedID` (which has a worse name xD)

**Review by vitorhugods — APPROVED** [review]

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/conversation/ConversationDAO.kt (hunk 3 above)
Comment (yamilmedina) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/conversation/ConversationDAO.kt:139]: It is to not introduce more variables to the new feature, we can Refactor later and unify once we are sure this is working as expected 🙂

**Pull request merged**